fixtures/_fixtures/monkeypatch.py

```diff
@@ -23,21 +23,91 @@
 from fixtures import Fixture
 
 
-def _setattr(obj, name, value):
-    """Handle some corner cases when calling setattr.
+class PartialClassMethod(object):
+    """Emulates the classmethod descriptor, but works on bound methods.
+
+    This wraps a function so that it acts like a classmethod on the calling
+    class. If the wrapped function is already bound to a class then the passed
+    parameters look like foo(boundcls_or_self, callingcls, *args, **kwargs).
 
-    setattr transforms a function into instancemethod, so where appropriate
-    value needs to be wrapped with staticmethod().
     """
+    def __init__(self, f):
+        self.f = f
+
+    def __call__(self, *args, **kwargs):
+        # Called when this is wrapped in another descriptor so __get__ is not
+        # used.
+        return self.f(*args, **kwargs)
+
+    def __get__(self, obj, objtype=None):
+        if objtype is None:
+            objtype = type(obj)
+
+        def newfunc(*args, **kwargs):
+            return self.f(objtype, *args, **kwargs)
+        return newfunc
```

Comment:
This is a great analysis.

The issue that led me to implement descriptors is that the other option I saw was to unpack a bound method and then create a partial function. But doing so meant determining whether to pull im_class or im_self in Python2 and I didn't even look into what Python3 has because I know im_self is dropped in favor of __self__ but I wasn't sure what the equivalent for im_class was since __class__ already has another meaning. The descriptor avoids touching any existing binding in place which means less to mess up.

I will cleanup to use partial here.